File: vulnassess/cvss31.py
"""CVSS v3.1 base and environmental scoring, implemented from the specification.

Pure arithmetic, no I/O and no third-party package.
"""

from math import floor

PREFIX = "CVSS:3.1"
BASE_METRICS = ("AV", "AC", "PR", "UI", "S", "C", "I", "A")
ORDER = (
    "AV", "AC", "PR", "UI", "S", "C", "I", "A",
    "E", "RL", "RC",
    "CR", "IR", "AR",
    "MAV", "MAC", "MPR", "MUI", "MS", "MC", "MI", "MA",
)

AV_VALUES = {"N": 0.85, "A": 0.62, "L": 0.55, "P": 0.20}
AC_VALUES = {"L": 0.77, "H": 0.44}
UI_VALUES = {"N": 0.85, "R": 0.62}
PR_UNCHANGED = {"N": 0.85, "L": 0.62, "H": 0.27}
PR_CHANGED = {"N": 0.85, "L": 0.68, "H": 0.50}
CIA_VALUES = {"H": 0.56, "L": 0.22, "N": 0.0}
REQUIREMENT_VALUES = {"H": 1.5, "M": 1.0, "L": 0.5, "X": 1.0}
E_VALUES = {"X": 1.0, "H": 1.0, "F": 0.97, "P": 0.94, "U": 0.91}
RL_VALUES = {"X": 1.0, "U": 1.0, "W": 0.97, "T": 0.96, "O": 0.95}
RC_VALUES = {"X": 1.0, "C": 1.0, "R": 0.96, "U": 0.92}

ALLOWED = {
    "AV": set(AV_VALUES), "AC": set(AC_VALUES), "PR": set(PR_UNCHANGED), "UI": set(UI_VALUES),
    "S": {"U", "C"}, "C": set(CIA_VALUES), "I": set(CIA_VALUES), "A": set(CIA_VALUES),
    "E": set(E_VALUES), "RL": set(RL_VALUES), "RC": set(RC_VALUES),
    "CR": set(REQUIREMENT_VALUES), "IR": set(REQUIREMENT_VALUES), "AR": set(REQUIREMENT_VALUES),
    "MAV": set(AV_VALUES) | {"X"}, "MAC": set(AC_VALUES) | {"X"},
    "MPR": set(PR_UNCHANGED) | {"X"}, "MUI": set(UI_VALUES) | {"X"},
    "MS": {"U", "C", "X"}, "MC": set(CIA_VALUES) | {"X"},
    "MI": set(CIA_VALUES) | {"X"}, "MA": set(CIA_VALUES) | {"X"},
}


def roundup(value: float) -> float:
    """The specification's round-half-up-to-one-decimal helper."""
    integer = int(round(value * 100000))
    if integer % 10000 == 0:
        return integer / 100000.0
    return (floor(integer / 10000) + 1) / 10.0


def parse(vector: str) -> dict[str, str]:
    """Parse a CVSS v3.x vector. A 4.0 vector is rejected: its arithmetic is different."""
    if not isinstance(vector, str) or not vector.startswith("CVSS:3."):
        raise ValueError(f"not a CVSS v3.x vector: {vector!r}")
    metrics: dict[str, str] = {}
    for token in vector.split("/")[1:]:
        if ":" not in token:
            raise ValueError(f"malformed metric {token!r} in {vector!r}")
        key, _, value = token.partition(":")
        if key not in ALLOWED:
            raise ValueError(f"unknown metric {key!r} in {vector!r}")
        if value not in ALLOWED[key]:
            raise ValueError(f"invalid value {key}:{value} in {vector!r}")
        metrics[key] = value
    missing = [key for key in BASE_METRICS if key not in metrics]
    if missing:
        raise ValueError(f"missing base metric(s) {missing} in {vector!r}")
    return metrics


def to_vector(metrics: dict[str, str]) -> str:
    parts = [f"{key}:{metrics[key]}" for key in ORDER if metrics.get(key) not in (None, "X")]
    return "/".join([PREFIX, *parts])


def _modified(metrics: dict[str, str], modified_key: str, base_key: str) -> str:
    value = metrics.get(modified_key, "X")
    return metrics[base_key] if value == "X" else value


def base_score(metrics: dict[str, str]) -> float:
    scope_changed = metrics["S"] == "C"
    iss = 1 - (
        (1 - CIA_VALUES[metrics["C"]])
        * (1 - CIA_VALUES[metrics["I"]])
        * (1 - CIA_VALUES[metrics["A"]])
    )
    if scope_changed:
        impact = 7.52 * (iss - 0.029) - 3.25 * (iss - 0.02) ** 15
    else:
        impact = 6.42 * iss
    privileges = (PR_CHANGED if scope_changed else PR_UNCHANGED)[metrics["PR"]]
    exploitability = (
        8.22
        * AV_VALUES[metrics["AV"]]
        * AC_VALUES[metrics["AC"]]
        * privileges
        * UI_VALUES[metrics["UI"]]
    )
    if impact <= 0:
        return 0.0
    if scope_changed:
        return roundup(min(1.08 * (impact + exploitability), 10))
    return roundup(min(impact + exploitability, 10))


def environmental_score(metrics: dict[str, str]) -> float:
    modified_scope_changed = _modified(metrics, "MS", "S") == "C"
    confidentiality = CIA_VALUES[_modified(metrics, "MC", "C")]
    integrity = CIA_VALUES[_modified(metrics, "MI", "I")]
    availability = CIA_VALUES[_modified(metrics, "MA", "A")]
    miss = min(
        1
        - (
            (1 - REQUIREMENT_VALUES[metrics.get("CR", "X")] * confidentiality)
            * (1 - REQUIREMENT_VALUES[metrics.get("IR", "X")] * integrity)
            * (1 - REQUIREMENT_VALUES[metrics.get("AR", "X")] * availability)
        ),
        0.915,
    )
    if modified_scope_changed:
        modified_impact = 7.52 * (miss - 0.029) - 3.25 * (miss * 0.9731 - 0.02) ** 13
    else:
        modified_impact = 6.42 * miss
    privileges = (PR_CHANGED if modified_scope_changed else PR_UNCHANGED)[
        _modified(metrics, "MPR", "PR")
    ]
    modified_exploitability = (
        8.22
        * AV_VALUES[_modified(metrics, "MAV", "AV")]
        * AC_VALUES[_modified(metrics, "MAC", "AC")]
        * privileges
        * UI_VALUES[_modified(metrics, "MUI", "UI")]
    )
    if modified_impact <= 0:
        return 0.0
    temporal = (
        E_VALUES[metrics.get("E", "X")]
        * RL_VALUES[metrics.get("RL", "X")]
        * RC_VALUES[metrics.get("RC", "X")]
    )
    combined = modified_impact + modified_exploitability
    if modified_scope_changed:
        return roundup(roundup(min(1.08 * combined, 10)) * temporal)
    return roundup(roundup(min(combined, 10)) * temporal)


def score_vector(vector: str) -> tuple[float, float]:
    metrics = parse(vector)
    return base_score(metrics), environmental_score(metrics)

File: vulnassess/ui/static/cvss31.js
const AV = {N: 0.85, A: 0.62, L: 0.55, P: 0.20};
const AC = {L: 0.77, H: 0.44};
const UI = {N: 0.85, R: 0.62};
const PR_U = {N: 0.85, L: 0.62, H: 0.27};
const PR_C = {N: 0.85, L: 0.68, H: 0.50};
const CIA = {H: 0.56, L: 0.22, N: 0};
const REQUIREMENTS = {H: 1.5, M: 1, L: 0.5, X: 1};
const EXPLOIT = {X: 1, H: 1, F: 0.97, P: 0.94, U: 0.91};
const REMEDIATION = {X: 1, U: 1, W: 0.97, T: 0.96, O: 0.95};
const CONFIDENCE = {X: 1, C: 1, R: 0.96, U: 0.92};
const ORDER = ['AV', 'AC', 'PR', 'UI', 'S', 'C', 'I', 'A', 'E', 'RL', 'RC', 'CR', 'IR', 'AR', 'MAV', 'MAC', 'MPR', 'MUI', 'MS', 'MC', 'MI', 'MA'];
const ALLOWED = {AV, AC, PR: PR_U, UI, S: {U: 1, C: 1}, C: CIA, I: CIA, A: CIA, E: EXPLOIT, RL: REMEDIATION, RC: CONFIDENCE, CR: REQUIREMENTS, IR: REQUIREMENTS, AR: REQUIREMENTS};
for (const [modified, base] of Object.entries({MAV: 'AV', MAC: 'AC', MPR: 'PR', MUI: 'UI', MS: 'S', MC: 'C', MI: 'I', MA: 'A'})) ALLOWED[modified] = {...ALLOWED[base], X: 1};

function pythonRound(value, digits) {
  if (!Number.isFinite(value)) throw new Error('A score must be finite.');
  if (value === 0) return value;
  const data = new DataView(new ArrayBuffer(8));
  data.setFloat64(0, Math.abs(value));
  const bits = data.getBigUint64(0);
  const encodedExponent = Number((bits >> 52n) & 2047n);
  const fraction = bits & ((1n << 52n) - 1n);
  const mantissa = encodedExponent ? fraction + (1n << 52n) : fraction;
  const exponent = (encodedExponent || 1) - 1023 - 52;
  let numerator = mantissa * 10n ** BigInt(digits);
  let denominator = 1n;
  if (exponent >= 0) numerator <<= BigInt(exponent);
  else denominator <<= BigInt(-exponent);
  let quotient = numerator / denominator;
  const remainder = numerator % denominator;
  if (remainder * 2n > denominator || (remainder * 2n === denominator && quotient % 2n === 1n)) quotient += 1n;
  return Math.sign(value) * Number(quotient) / 10 ** digits;
}

function roundup(value) {
  const integer = pythonRound(value * 100000, 0);
  return integer % 10000 === 0 ? integer / 100000 : (Math.floor(integer / 10000) + 1) / 10;
}

export function parseVector(vector) {
  if (typeof vector !== 'string' || !vector.startsWith('CVSS:3.')) throw new Error('A stored CVSS 3.x vector is required.');
  const metrics = {};
  for (const token of vector.split('/').slice(1)) {
    const [key, value, extra] = token.split(':');
    if (extra !== undefined || !Object.hasOwn(ALLOWED, key) || !Object.hasOwn(ALLOWED[key], value)) throw new Error(`Invalid CVSS metric: ${token}`);
    metrics[key] = value;
  }
  for (const key of ORDER.slice(0, 8)) if (!Object.hasOwn(metrics, key)) throw new Error(`Missing CVSS metric: ${key}`);
  return metrics;
}

function modified(metrics, name, base) {
  const value = metrics[name] ?? 'X';
  return value === 'X' ? metrics[base] : value;
}

function baseScore(metrics) {
  const changed = metrics.S === 'C';
  const impactSubscore = 1 - ((1 - CIA[metrics.C]) * (1 - CIA[metrics.I]) * (1 - CIA[metrics.A]));
  const impact = changed ? 7.52 * (impactSubscore - 0.029) - 3.25 * (impactSubscore - 0.02) ** 15 : 6.42 * impactSubscore;
  const privileges = (changed ? PR_C : PR_U)[metrics.PR];
  const exploitability = 8.22 * AV[metrics.AV] * AC[metrics.AC] * privileges * UI[metrics.UI];
  if (impact <= 0) return 0;
  return roundup(Math.min((changed ? 1.08 : 1) * (impact + exploitability), 10));
}

function environmentalScore(metrics) {
  const changed = modified(metrics, 'MS', 'S') === 'C';
  const confidentiality = CIA[modified(metrics, 'MC', 'C')];
  const integrity = CIA[modified(metrics, 'MI', 'I')];
  const availability = CIA[modified(metrics, 'MA', 'A')];
  const impactSubscore = Math.min(1 - (
    (1 - REQUIREMENTS[metrics.CR ?? 'X'] * confidentiality) *
    (1 - REQUIREMENTS[metrics.IR ?? 'X'] * integrity) *
    (1 - REQUIREMENTS[metrics.AR ?? 'X'] * availability)
  ), 0.915);
  const impact = changed ? 7.52 * (impactSubscore - 0.029) - 3.25 * (impactSubscore * 0.9731 - 0.02) ** 13 : 6.42 * impactSubscore;
  const privileges = (changed ? PR_C : PR_U)[modified(metrics, 'MPR', 'PR')];
  const exploitability = 8.22 * AV[modified(metrics, 'MAV', 'AV')] * AC[modified(metrics, 'MAC', 'AC')] * privileges * UI[modified(metrics, 'MUI', 'UI')];
  if (impact <= 0) return 0;
  const temporal = EXPLOIT[metrics.E ?? 'X'] * REMEDIATION[metrics.RL ?? 'X'] * CONFIDENCE[metrics.RC ?? 'X'];
  return roundup(roundup(Math.min((changed ? 1.08 : 1) * (impact + exploitability), 10)) * temporal);
}

export function scoreVector(vector) {
  const metrics = parseVector(vector);
  return {base: baseScore(metrics), environmental: environmentalScore(metrics)};
}

export function sandbox(vector, profile, changes, weights) {
  const metrics = parseVector(vector);
  const settings = weights.environmental;
  const minimum = settings.min_confidence_to_apply ?? 0.5;
  const local = JSON.parse(JSON.stringify(profile));
  const feature = value => ({value, confidence: 1, source: 'manual', evidence: 'Sandbox assumption'});
  local.controls ||= {};
  local.manual ||= {};
  if (Object.hasOwn(changes, 'role')) local.role = feature(changes.role);
  if (Object.hasOwn(changes, 'exposure')) local.exposure = feature(changes.exposure);
  if (Object.hasOwn(changes, 'waf')) local.controls.waf = feature(changes.waf);
  if (Object.hasOwn(changes, 'environment')) local.manual.environment = feature(changes.environment);
  if (local.exposure.value === 'internal' && metrics.AV === 'N' && local.exposure.confidence >= minimum) Object.assign(metrics, settings.internal_when_av_network);
  if (local.controls.waf?.value && local.controls.waf.confidence >= minimum) Object.assign(metrics, settings.waf_present);
  if (local.controls.auth_required?.value && metrics.PR === 'N' && local.controls.auth_required.confidence >= minimum) Object.assign(metrics, settings.auth_required_when_pr_none);
  const role = local.role.confidence >= minimum ? local.role.value : 'unknown';
  let requirements = {...(settings.role_requirements[role] || settings.role_requirements.unknown)};
  if (local.manual.environment?.value === 'test') requirements = {...settings.environment_test};
  const steps = Math.max(0, Number(local.manual.criticality?.value || 0) - settings.criticality_step_above);
  const stepUp = {L: 'M', M: 'H', H: 'H'};
  for (let step = 0; step < steps; step += 1) requirements = Object.fromEntries(Object.entries(requirements).map(([key, value]) => [key, stepUp[value]]));
  Object.assign(metrics, requirements);
  const environmental = environmentalScore(metrics);
  const percentile = changes.epss;
  if (percentile !== null && (!Number.isFinite(percentile) || percentile < 0 || percentile > 1)) throw new Error('EPSS percentile must be in 0..1 or absent.');
  let multiplier = percentile === null ? null : weights.threat.base_multiplier + weights.threat.epss_weight * percentile;
  if (changes.kev) multiplier = Math.max(multiplier || 0, weights.threat.kev_multiplier);
  let risk = multiplier === null ? environmental * 10 : environmental * 10 * multiplier;
  if (changes.kev) {
    risk += weights.threat.kev_boost;
    if (local.exposure.value === 'internet_facing') risk = Math.max(risk, weights.threat.kev_floor_internet_facing);
  }
  risk = pythonRound(Math.min(risk, 100), 1);
  const band = risk >= weights.bands.Critical ? 'Critical' : risk >= weights.bands.High ? 'High' : risk >= weights.bands.Medium ? 'Medium' : 'Low';
  return {
    base: baseScore(metrics), environmental, risk, band, multiplier,
    vector: ['CVSS:3.1', ...ORDER.filter(key => metrics[key] && metrics[key] !== 'X').map(key => `${key}:${metrics[key]}`)].join('/')
  };
}